File: lab-1/src/main/java/com/evist0/labforwardlist/factory/ObjectFactoryRegistry.java
package com.evist0.labforwardlist.factory;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.ServiceLoader;
import java.util.stream.Collectors;

public abstract class ObjectFactoryRegistry {
    private static final List<ObjectFactory> factories;

    static {
        factories = new ArrayList<>();

        ServiceLoader<ObjectFactory> loadedFactories = ServiceLoader.load(ObjectFactory.class);

        for (ObjectFactory factory : loadedFactories) {
            factories.add(factory);
        }
    }

    public static ObjectFactory getFactory(String factoryName) {
        Objects.requireNonNull(factoryName);

        return factories.stream().filter(x -> x.getObjectName().equals(factoryName)).findFirst().orElse(null);
    }

    public static List<String> getAllFactories() {
        return factories.stream().map(ObjectFactory::getObjectName).collect(Collectors.toList());
    }
}

File: lab-2/src/main/scala/com/hamaddo/labforwardlist/factory/ObjectFactoryRegistry.scala
package com.hamaddo.labforwardlist.factory

import java.util
import java.util.stream.Collectors
import java.util.{Objects, ServiceLoader}

object ObjectFactoryRegistry {
  private val factories = new util.ArrayList[ObjectFactory]()

  private val loadedFactories: ServiceLoader[ObjectFactory] = ServiceLoader.load(classOf[ObjectFactory])

  loadedFactories.forEach(factory => {
    factories.add(factory)
  })

  def getFactory(factoryName: String): ObjectFactory = {
    Objects.requireNonNull(factoryName)

    factories.stream().filter(x => x.getObjectName.equals(factoryName)).findFirst().orElse(null)
  }

  def getAllFactories: util.List[String] = {
    factories.stream().map(x => x.getObjectName).collect(Collectors.toList[String])
  }
}
